doc/designs/application-specific-passwords.md

```diff
@@ -0,0 +1,283 @@
+
+~~As both of these authentication mechanisms include sending a password for
+authentication, they could also be used with application-specific passwords.
+As pure RADIUS provides only weak protection of the sent password, it might
+be even more desirable than in the case of traditional password
+authentication.~~
+
+## Design
+
+### Multiple passwords for a single user and related changes
+
+Application-specific passwords will be stored in a new LDAP subtree in the form
+```
+object,cn=appname,cn=apps,$SUFFIX.
+```
+Every application will have its own subtree in which users will be able to
+create their objects if they will want to use that application - this means
+objects containing application-specific passwords will not be grouped by users,
+but by applications. There will be one object for each application-specific
+password of each user; one user will be able to have one password for a given
+application. The object containing an application-specific password will have
+the following attributes:
+* `krbprincipalkey` - will hold the application-specific password itself
+* `krbprincipalname` - two-component Kerberos principal (see below)
+* backlink to the owning user (that created it) - this backlink must be
+  maintained; inspiration can be seen in working with hostgroups
+
+The definition of the new object and its attributes holding application-specific
+passwords and related information, as well as operations with them (add, modify,
+delete) will be contained in a new file - `ipaserver/plugins/appspecificpw.py`.
+
+Names of the newly created objects in the new subtree will somehow have to be
+validated and enforced (possibly, but not necessarily using an LDAP plugin), so
+that the user will not be able to create a principal looking like a host or a
+service, or even set a password for another user.
+
+The existing mechanism for generating passwords will be made available for use
+for application-specific passwords as well. Of course, the user will be able to
+set application-specific passwords as usual (having to fulfil password policies
+in place).
+
+Password reset functionality must be extended to facilitate multiple passwords
+for a single user as well - the user will be able to specify what password to
+reset using the application name of the application-specific password. If it
+will be left blank, then primary password will be reset, otherwise
+application-specific password for the respective application will be reset. A
+new parameter therefore needs to be added to the password reset command -
+application name (for the application-specific password).
+
+The user account status will display information for all its passwords.
+
+### ipa-kdb support for the aforementioned functionality
+
+The ipa-kdb daemon will be extended to support application-specific passwords
+as well, so that they can be used with Kerberos and HBAC.
+
+Subaccounts will be introduced with principals in the form 
+```_appname_/_user_@_REALM_```
```

Comment:
This is backwards traditionally two component principals are: username/instance@REALM
We MUST NOT deviate from this at will create useless confusion among admins that already understand kerberos and instance specific principals, as well as cause issue in configuring applications as all instructions for instance specific configuration on the web today will reference that model.

Reply:
All right, so it will be user/appname@REALM, as I originally assumed, right? Just to make clear, the underscores in the design page is a GitHub's markup for itallic font - I did this to emphasize that those terms in the principal are not constants, but variables.